warden_daemon/src/socket/vsocket_server.rs

```diff
@@ -0,0 +1,132 @@
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use futures::SinkExt;
+use log::{info, trace};
+use thiserror::Error;
+use tokio::select;
+use tokio::sync::oneshot::{Receiver, Sender};
+use tokio::sync::{oneshot, Mutex};
+use tokio_serde::formats::SymmetricalJson;
+use tokio_util::codec::{FramedWrite, LengthDelimitedCodec};
+use tokio_util::sync::CancellationToken;
+use tokio_vsock::{VsockAddr, VsockListener, VsockStream};
+
+use crate::client_handler::realm_client_handler::{
+    RealmCommand, RealmConnector, RealmSender, RealmSenderError,
+};
+
+pub struct VSockServerConfig {
+    pub cid: u32,
+    pub port: u32,
+}
+
+#[derive(Debug, Error)]
+pub enum VSockServerError {
+    #[error("Error while sending RealmSender through the channel.")]
+    ChannelFail,
+    #[error("Unknown Realm has connected.")]
+    UnexpectedConnection,
+    #[error("Socket failure has occured: {0}")]
+    SocketFail(#[from] io::Error),
+}
+
+pub struct VSockServer {
+    config: VSockServerConfig,
+    waiting: HashMap<u32, Sender<Box<dyn RealmSender + Send + Sync>>>,
+}
+
+impl VSockServer {
+    pub fn new(config: VSockServerConfig) -> Self {
+        VSockServer {
+            config,
+            waiting: HashMap::new(),
+        }
+    }
+
+    pub async fn listen(
+        handler: Arc<Mutex<VSockServer>>,
+        token: Arc<CancellationToken>,
+    ) -> Result<(), VSockServerError> {
+        let mut listener = {
+            let config = &handler.as_ref().lock().await.config;
+            VsockListener::bind(VsockAddr::new(config.cid, config.port))
+                .map_err(VSockServerError::SocketFail)?
+        };
+        loop {
+            select! {
+                a_result = listener.accept() => {
+                    match a_result {
+                        Ok(result) => {
+                            trace!("Accepted connection.");
+                            let mut handler = handler.lock().await;
+                            handler.handle_accept(result).await?
+                        },
+                        Err(err) => return Err(VSockServerError::SocketFail(err))
+                    }
+                }
+                _ = token.cancelled() => {return Ok(());}
+
+            }
+        }
+    }
+
+    async fn handle_accept(
+        &mut self,
+        accept_result: (VsockStream, VsockAddr),
+    ) -> Result<(), VSockServerError> {
+        let (stream, addr) = accept_result;
+        if let Some(tx) = self.waiting.remove(&addr.cid()) {
+            info!("Client has connected succesfully.");
+            return tx
+                .send(Box::new(VSockClient::new(stream)))
+                .map_err(|_| VSockServerError::ChannelFail);
+        }
+        Err(VSockServerError::UnexpectedConnection)
+    }
+}
+
+struct VSockClient {
+    stream: tokio_serde::Framed<
+        FramedWrite<VsockStream, LengthDelimitedCodec>,
+        RealmCommand,
+        RealmCommand,
+        tokio_serde::formats::Json<RealmCommand, RealmCommand>,
```

Comment:
Sure, I just wanted to make it easier now for testing purposes because in the future we will establish a protocol between Realm and Warden and this part of a code will be changed.